bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,360 @@
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem == null, "HDFSStorage has already been initialized.");
+        Configuration conf = new Configuration();
+        conf.set("fs.default.name", this.config.getHdfsHostURL());
+        conf.set("fs.default.fs", this.config.getHdfsHostURL());
+        conf.set("fs.hdfs.impl", "org.apache.hadoop.hdfs.DistributedFileSystem");
+
+        // FileSystem has a bad habit of caching Clients/Instances based on target URI. We do not like this, since we
+        // want to own our implementation so that when we close it, we don't interfere with others.
+        conf.set("fs.hdfs.impl.disable.cache", "true");
+        if (!this.config.isReplaceDataNodesOnFailure()) {
+            // Default is DEFAULT, so we only set this if we want it disabled.
+            conf.set("dfs.client.block.write.replace-datanode-on-failure.policy", "NEVER");
+        }
+
+        this.fileSystem = openFileSystem(conf);
+        log.info("Initialized (HDFSHost = '{}'", this.config.getHdfsHostURL());
+    }
+
+    private FileSystem openFileSystem(Configuration conf) throws IOException {
+        return FileSystem.get(conf);
+    }
+    //endregion
+
+    //region Helpers
+    private void ensureInitializedAndNotClosed() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem != null, "HDFSStorage is not initialized.");
+    }
+
+    //endregion
+
+    //Region HDFS helper methods.
+
+    /**
+     * Gets an HDFS-friendly path prefix for the given chunk name by pre-pending the HDFS root from the config.
+     */
+    private String getPathPrefix(String chunkName) {
+        return this.config.getHdfsRoot() + Path.SEPARATOR + chunkName;
+    }
+
+    /**
+     * Gets the full HDFS Path to a file for the given chunk, startOffset and epoch.
+     */
+    private Path getFilePath(String chunkName) {
+        Preconditions.checkState(chunkName != null && chunkName.length() > 0, "chunkName must be non-null and non-empty");
+        return new Path(getPathPrefix(chunkName));
+    }
+
+    /**
+     * Determines whether the given FileStatus indicates the file is read-only.
+     *
+     * @param fs The FileStatus to check.
+     * @return True or false.
+     */
+    private boolean isReadOnly(FileStatus fs) {
+        return fs.getPermission().getUserAction() == FsAction.READ;
+    }
+
+    private int readInternal(ChunkHandle handle, byte[] buffer, long offset, int bufferOffset, int length) throws IOException {
+        //There is only one file per chunkName.
+        try (FSDataInputStream stream = this.fileSystem.open(getFilePath(handle.getChunkName()))) {
+            stream.readFully(offset, buffer, bufferOffset, length);
+        } catch (EOFException e) {
+            throw new IllegalArgumentException(String.format("Reading at offset (%d) which is beyond the current size of chunk.", offset));
```

Comment:
Could we check this as a precondition?

Reply:
Precondition is checked elsewhere. But still need to handle runtime exception.